src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,403 @@
+	skip_mixed_frames = mixed_frames - start_frame;
+
+	if (frame_count <= skip_mixed_frames)
+		return;
+	frame_count -= skip_mixed_frames;
+	channel_count = stream->channels;
+	/* audio_stream_wrap() is needed here and it is just below in a loop */
+	ptr = (int16_t *)stream->w_ptr + mixed_frames * stream->channels + channel_index;
+
+	for (left_frames = frame_count; left_frames; left_frames -= frames) {
+		ptr = audio_stream_wrap(stream, ptr);
+		n = audio_stream_samples_without_wrap_s16(stream, ptr);
+		n =  MIN(left_frames * channel_count, n);
+		frames = 0;
+		for (i = 0; i < n; i += channel_count) {
+			*ptr = 0;
+			ptr += channel_count;
+			frames++;
+		}
+	}
+}
+#endif	/* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function
+ * to also mix an entire stream. In this case the function is called with fake stream
+ * parameters: multichannel stream is treated as single channel and so the entire stream
+ * contents is mixed.
+ */
+static void normal_mix_channel_s24(struct audio_stream __sparse_cache *sink, uint32_t start_frame,
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+	/* audio_stream_wrap() is required and is done below in a loop */
+	int32_t *dst = (int32_t *)sink->w_ptr + start_frame;
+	int32_t *src = (int32_t *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s24(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = sat_int24(sign_extend_s24(*dst) + sign_extend_s24(*src++));
+			dst++;
+		}
+	}
+
+	for (left_frames = frames_to_copy; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		nmax = audio_stream_samples_without_wrap_s24(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++)
+			*dst++ = *src++;
+	}
+}
+
+static void remap_mix_channel_s24(struct audio_stream __sparse_cache *sink,
+				  uint8_t sink_channel_index,  uint8_t sink_channel_count,
+				  uint32_t start_frame, uint32_t mixed_frames,
+				  const struct audio_stream __sparse_cache *source,
+				  uint8_t source_channel_index, uint8_t source_channel_count,
+				  uint32_t frame_count, uint16_t gain)
```

Comment:
Here the same uint8_t parameters galore issue. Struct with int for generic numbers?

Reply:
change uint8_t to int32_t can have save about 0.6% cycles in remap_mix_channel_s16, so I will change the parameters to int32_t, but not struct.

Reply:
OK, but be careful when mixing unsigned and signed arithmetic. It's safest to use only signed types. It's very rare that we need to overflow the integers (a stupid shortcoming in C standard to claim it's undefined) or need the 2x positive range of unsigned.